elm/util/diff_eval.py

```diff
@@ -0,0 +1,256 @@
+    # If we use the line numbers, we match-and-replace in a line-by-line fashion.
+    file_by_line = file.split('\n') if use_line_number else None
+    line_offset = 0  # the offset between pre-/post-patching line numbers
+
+    # If we do not use the line numbers, for multiple diff hunk, we only move forward in a greedy manner.
+    patch_pointer = 0
+
+    i = 0 if diff[0] else 1  # We have delimiter at the beginning, causing empty initial string
+    while i < len(diff) - 1:  # Need at least a pair of '@@ ... @@' and diff hunk to continue
+        # Expect a string with '@@ ... @@' followed by a diff hunk
+        line_info = parse_line_info(diff[i])
+        diff_content = diff[i + 1]
+        i += 2
+
+        # Generate the pre-/post-diff string based on the first character being '+' or '-'
+        # (Note: parse_diff_content will strip trailing \n at the beginning and the end)
+        parsed_diff = parse_diff_content(diff_content, separate_lines=use_line_number)
+
+        # If we allow the recognition of "ADDFILE" and encounter such file, special treatment is needed.
+        if allow_add_file and file == "ADDFILE":
+            if use_line_number:
+                return parsed_diff[1] if line_info == (0, 0) else ""
+            else:
+                return parsed_diff[1]
+
+        if use_line_number:
+            # If line numbers cannot be parsed, skip.
+            if not line_info:
+                continue
+
+            # Offset the starting line
+            start_idx = line_info[0] + line_offset
+
+            # Match the referred lines with the file context
+            referred_lines = file_by_line[start_idx - 1 : start_idx - 1 + line_info[1]]
+            valid = all([l1 == l2 for l1, l2 in zip(parsed_diff[0], referred_lines)])
+
+            # If lines fully match and the number of lines is consistent, apply the patch.
+            # We ignore the second pair "+a, b" just to be lenient.
+            if valid and len(parsed_diff[0]) == line_info[1]:
+                # Update the list of lines
+                file_by_line = file_by_line[: start_idx - 1] + parsed_diff[1] + \
+                               file_by_line[start_idx - 1 + line_info[1] :]
+                line_offset += len(parsed_diff[1]) - line_info[1]
+        else:
+            # Directly (and naively) apply patch by match-and-replace.
+            file, patch_pointer = replace_text(file, parsed_diff[0], parsed_diff[1], patch_pointer)
+
+    if use_line_number:
+        file = "\n".join(file_by_line)
+    return file
+
+
+def verify_diff(diff_text: str) -> int:
+    """
+    Verify the validity of a complete diff text.
+
+    Args:
+        diff_text: the complete diff text.
+            The overall format conforms "<NME> ...\n<BEF> ...\n<MSG> ...\n<DFF> ..." and the text
+            after <DFF> has 1 or more lines of "@@ -x,y +a,b @@" followed by the corresponding hunk.
+    Returns:
+        a custom status code (binary):
+            0b000: valid diff.
```

Comment:
Can we make the status code an Enum? Just seems a bit neater maybe?

Reply:
Yeah that's a good point!